### PR TITLE
refactor: use ternary when creating RPC param variables

### PR DESCRIPTION
**Pull request by fanquake**
A style we are already using in some RPC code. i.e:
https://github.com/bitcoin/bitcoin/blob/55ea6fd2506b84f553c0bd42f48a9cc4393bab47/src/rpc/blockchain.cpp#L864-L866
More concise (about half the lines of code).
Variables that should be made `const`, can be made `const`.
I'd argue that this:
```cpp
    const bool fVerbose{request.params[0].isNull() ? false : request.params[0].get_bool()};

    const bool include_mempool_sequence{request.params[1].isNull() ? false : request.params[1].get_bool()};
```
is easier to read / reason about, than:
```cpp
    bool fVerbose = false;
    if (!request.params[0].isNull())
        fVerbose = request.params[0].get_bool();

    bool include_mempool_sequence = false;
    if (!request.params[1].isNull()) {
        include_mempool_sequence = request.params[1].get_bool();
    }
```

### CODE REVIEW

**Comment by jonatack** [comment]
Concept ACK. I've been suggesting this style in code reviews since a year or two.

**Comment by prusnak** [comment]
What do you think about further removing the boilerplate code using macros?

e.g. the following

``` c++
const bool fSubtractFeeFromAmount{request.params[4].isNull() ? false : request.params[4].get_bool()};
const int timeout {request.params[1].isNull() ? 0 : request.params[1].get_int()};
const double threshold{request.params[1].isNull() ? 0.95 : request.params[1].get_real()};
const std::string strCommand{request.params[1].isNull() ? "" : request.params[1].get_str()};
```

becomes:

``` c++
#define BOOL_ARG(i, name, defval) const bool #name {request.params[i].isNull() ? defval : request.params[i].get_bool()}
#define INT_ARG(i, name, defval) const int #name {request.params[u].isNull() ? defval : request.params[i].get_int()}
#define DOUBLE_ARG(i, name, defval) const double #name {request.params[i].isNull() ? defval : request.params[i].get_real()}
#define STRING_ARG(i, name, defval) const std::string #name {request.params[i].isNull() ? defval : request.params[i].get_str()}

BOOL_ARG(4, fSubtractFeeFromAmount, false);
INT_ARG(1, timeout, 0);
DOUBLE_ARG(1, threshold, 0.95);
STRING_ARG(1, strCommand, "");
```

**Comment by ajtowns** [comment]
Could just be an inline method for JSONRPCRequest?

```c++
class JSONRPCRequest
{
    ...
    bool get_bool_param(size_t id, bool default) const
    {
        return params[id].isNull() ? default : params[id].get_bool();
    }
};
fSubtractFeeFromAmount{request.get_bool_param(4, false)};
```

**Comment by prusnak** [comment]
> Could just be an inline method for JSONRPCRequest?

Will this still allow creating a const variable?

File: src/rpc/blockchain.cpp [hunk 1]
```diff
@@ -359,13 +354,10 @@ static RPCHelpMan waitforblockheight()
                 },
         [&](const RPCHelpMan& self, const JSONRPCRequest& request) -> UniValue
 {
-    int timeout = 0;
+    const int timeout {request.params[1].isNull() ? 0 : request.params[1].get_int()};
```
Comment (shaavan) [inline src/rpc/blockchain.cpp:357]: nit:
```suggestion
    const int timeout{request.params[1].isNull() ? 0 : request.params[1].get_int()};
```

File: src/rpc/blockchain.cpp [hunk 2]
```diff
@@ -2191,10 +2179,7 @@ static RPCHelpMan getblockfilter()
         [&](const RPCHelpMan& self, const JSONRPCRequest& request) -> UniValue
 {
     uint256 block_hash = ParseHashV(request.params[0], "blockhash");
-    std::string filtertype_name = "basic";
-    if (!request.params[1].isNull()) {
-        filtertype_name = request.params[1].get_str();
-    }
+    const std::string filtertype_name = {request.params[1].isNull() ? "basic" : request.params[1].get_str()};
```
Comment (shaavan) [inline src/rpc/blockchain.cpp:2182]: nit:
```suggestion
    const std::string filtertype_name{request.params[1].isNull() ? "basic" : request.params[1].get_str()};
```

**Review by shaavan — COMMENTED** [review]
Concept ACK

- I like the idea of using the ternary operator to define param variables, as this makes the code clean and easy to understand and reason with. However, I see some excellent update ideas to this PR, and I would be interested to see how they would be taken.

In the meantime, I found some nits that would make the changes in this PR consistent.

**Review by promag — COMMENTED** [review]
Concept ACK.

FWIW #20017 takes a different approach to handle RPC arguments.

**Comment by MarcoFalke** [comment]
@promag Heh, was about to link that pull just now

**Comment by laanwj** [comment]
I don't like these kind of "change style all over the place for sake of changing style" PRs to be honest. That said I think using the tertiary operator is fine, as long as it's only one level deep. It can become nasty and hard to review when nested.

**Comment by MarcoFalke** [comment]
Agree with @laanwj . Ideally we should make those changes so that they give additional run-time or compile-time guarantees. Otherwise, the lines that are changed here will need to be touched again if default arguments are made "documentation-enforced".

**Comment by fanquake** [comment]
> I don't like these kind of "change style all over the place for sake of changing style" PRs to be honest. 

> Ideally we should make those changes so that they give additional run-time or compile-time guarantees.

I think I want to push back against "style for the sake of style" here. It's not often that style-only changes result in half as many lines of (I'll claim somewhat "better") code. IIRC the binary for this branch is also smaller.

I realise there is a spectrum, but I think at some point we have to allow these kind of "cleanup" PRs, that may be slightly closer to stylish than we'd like, because if we do, it should eventually just result in "better", and maybe generally less, code in the repository.

Saying all that, it seems like something like #20017 is probably the better way forward here.

**Pull request closed without merging**